hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCopyTest.java

```diff
@@ -0,0 +1,162 @@
+import static org.apache.hadoop.fs.contract.ContractTestUtils.verifyFileContents;
+import static org.apache.hadoop.test.LambdaTestUtils.intercept;
+import static org.apache.hadoop.test.LambdaTestUtils.interceptFuture;
+
+/**
+ * Test native copy - if supported.
+ */
+public abstract class AbstractContractCopyTest extends
+    AbstractFSContractTestBase {
+
+  protected Path srcDir;
+  protected Path srcFile;
+  protected Path zeroByteFile;
+  protected Path dstDir;
+  protected Path dstFile;
+  protected FileSystem fs;
+
+  protected byte[] srcData;
+
+  @Override
+  public void setup() throws Exception {
+    super.setup();
+    skipIfUnsupported(SUPPORTS_NATIVE_COPY);
+
+    srcDir = absolutepath("source");
+    dstDir = absolutepath("target");
+
+    srcFile = new Path(srcDir, "source.txt");
+    dstFile = new Path(dstDir, "dst.txt");
+    zeroByteFile = new Path(srcDir, "zero.txt");
+
+    srcData = dataset(TEST_FILE_LEN, 0, 255);
+    createFile(getFileSystem(), srcFile, true, srcData);
+    touch(getFileSystem(), zeroByteFile);
+
+    fs = getFileSystem();
+  }
+
+  @Override
+  public void teardown() throws Exception {
+    describe("\nTeardown\n");
+    super.teardown();
+  }
+
+
+  @Test
+  public void testBasicCopy() throws Throwable {
+    describe("Checks whether S3A FS supports native copy"
+        + " and copies a file from one location to another."
+        + " File deletion taken care in teardown.");
+
+    assertPathExists(srcFile + " does not exist", srcFile);
+
+    fs.delete(dstFile, false);
+    // initiate copy
+    fs.copyFile(srcFile.toUri(), dstFile.toUri()).get();
+
+    // Check if file is available and verify its contents
+    assertIsFile(srcFile);
+    assertIsFile(dstFile);
+    verifyFileContents(fs, dstFile, srcData);
+  }
+
+  @Test
+  public void testZeroByteFile() throws Throwable {
+    describe("Copy 0 byte file.");
+
+    assertPathExists(zeroByteFile + " does not exist", zeroByteFile);
+
+    fs.delete(dstFile, false);
+    // initiate copy
+    fs.copyFile(zeroByteFile.toUri(), dstFile.toUri()).get();
+    // Check if file is available
+    assertIsFile(zeroByteFile);
+  }
+
+  @Test
+  public void testFolderCopy() throws Throwable {
+    describe("Test folder copy. Recursive copy is not supported yet.");
+    // initiate copy
+    handleExpectedException(intercept(Exception.class,
+        () -> fs.copyFile(srcDir.toUri(), dstDir.toUri()).get()));
+  }
+
+  @Test
+  public void testOverwrite() throws Throwable {
+    describe("Test copying a file which is already present");
+
+    fs.delete(dstFile, false);
+    // initiate copy
+    fs.copyFile(srcFile.toUri(), dstFile.toUri()).get();
+
+    // Recopy, this would overwrite the contents
+    fs.copyFile(srcFile.toUri(), dstFile.toUri()).get();
+
+    // Check if file is available and verify its contents
+    assertIsFile(srcFile);
+    assertIsFile(dstFile);
+    verifyFileContents(fs, dstFile, srcData);
```

Comment:
for s3 we need to spin a bit here for eventual consistency; use LambdaTestUtils.eventually()